scripts/launch/demo-cluster/default.nix

```diff
@@ -0,0 +1,128 @@
+{ localLib ? import ./../../../lib.nix
+, stateDir ? localLib.maybeEnv "CARDANO_STATE_DIR" "./state-demo"
+, config ? {}
+, runWallet ? true
+, runExplorer ? false
+, numCoreNodes ? 4
+, system ? builtins.currentSystem
+, pkgs ? import localLib.fetchNixPkgs { inherit system config; }
+, gitrev ? localLib.commitIdFromGitRepo ./../../../.git
+, ghcRuntimeArgs ? "-N2 -qg -A1m -I0 -T"
+, additionalNodeArgs ? ""
+, keepAlive ? true
+}:
+
+with localLib;
+
+let
+  executables =  {
+    corenode = "${iohkPkgs.cardano-sl-node-static}/bin/cardano-node-simple";
+    wallet = "${iohkPkgs.cardano-sl-wallet-new}/bin/cardano-node";
+    integration-test = "${iohkPkgs.cardano-sl-wallet-new}/bin/cardano-integration-test";
+    keygen = "${iohkPkgs.cardano-sl-tools}/bin/cardano-keygen";
+    explorer = "${iohkPkgs.cardano-sl-explorer-static}/bin/cardano-explorer";
+  };
+  ifWallet = localLib.optionalString (runWallet);
+  ifKeepAlive = localLib.optionalString (keepAlive);
+  iohkPkgs = import ./../../.. { inherit config system pkgs gitrev; };
+  src = ./../../..;
+  configFiles = pkgs.runCommand "cardano-config" {} ''
+    mkdir -pv $out
+    cd $out
+    cp -vi ${iohkPkgs.cardano-sl.src + "/configuration.yaml"} configuration.yaml
+    cp -vi ${iohkPkgs.cardano-sl.src + "/mainnet-genesis-dryrun-with-stakeholders.json"} mainnet-genesis-dryrun-with-stakeholders.json
+    cp -vi ${iohkPkgs.cardano-sl.src + "/mainnet-genesis.json"} mainnet-genesis.json
+  '';
+in pkgs.writeScript "demo-cluster" ''
+  #!${pkgs.stdenv.shell}
+  export PATH=${pkgs.lib.makeBinPath (with pkgs; (with iohkPkgs; [ jq coreutils pkgs.curl gnused openssl cardano-sl-tools cardano-sl-wallet-new cardano-sl-node-static ]))}
+  EXIT_STATUS=0
+  source ${src + "/scripts/common-functions.sh"}
+  LOG_TEMPLATE=${src + "/log-configs/template-demo.yaml"}
+  function stop_cardano {
+    trap "" INT TERM
+    echo "Received TERM!"
+    echo "Stopping Cardano core nodes"
+    for pid in ''${core_pid[@]}
+    do
+      echo killing pid $pid
+      kill $pid
+    done
+    ${ifWallet ''
+      echo killing wallet pid $wallet_pid
+    kill $wallet_pid
+    ''}
+    wait
+    echo "Stopped all Cardano processes, exiting!"
+    exit $EXIT_STATUS
+  }
+  system_start=$((`date +%s` + 15))
+  echo "Using system start time "$system_start
+
+
+
+  # Remove previous state
+  rm -rf ${stateDir}
+  mkdir -p ${stateDir}
+  echo "Creating genesis keys..."
+  cardano-keygen --system-start 0 generate-keys-by-spec --genesis-out-dir ${stateDir}/genesis-keys --configuration-file ${configFiles}/configuration.yaml
+
+  echo "Generating Topology"
+  gen_kademlia_topology ${builtins.toString (numCoreNodes + 1)} ${stateDir}
```

Comment:
does this need the `+ 1`? The comment to the function says it will generate `$1` files.

Reply:
Yes, in this case, the wallet has to be in the topology to talk to the core nodes.